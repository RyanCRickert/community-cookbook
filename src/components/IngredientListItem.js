import React from "react";

export default (props) => (
  <div className="ingredient-item">
    <div className="ingredient-item__left">
      {props.name}
    </div>
    <div
      className="ingredient-item__right"
      onClick={(e) => {
        props.handleRemoveIngredient(props.name)
      }}>
      Remove   
    </div>
  </div>
)