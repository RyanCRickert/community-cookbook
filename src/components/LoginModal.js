import React from "react";
import Modal from "react-modal";
import { connect } from "react-redux";
import { firebase, googleAuthProvider } from "../firebase/firebase";
import { startGoogleLogin } from "../actions/auth";

export class LoginModal extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			email: "",
			password: "",
			newUser: false,
			error: ""
      
		}
	}

	onPasswordChange = (e) => {
		const password = e.target.value;
		this.setState(() => ({ password }));
	};

	onEmailChange = (e) => {
		const email = e.target.value;
    this.setState(() => ({ email }));
	};

	handleLogin = (e) => {
		e.preventDefault();

		firebase.auth().signInWithEmailAndPassword(this.state.email, this.state.password).catch((e) => {
			if(e.code === "auth/user-not-found") {
				this.setState({ error: "No user with that email found." });
			} else if (e.code === "auth/wrong-password") {
				this.setState({ error: "Incorrect password, please try again." });
			} else {
				this.setState({ error: e.message });
			}
    })
	};

	handleStartNewUser = () => {
		this.setState({ newUser: true });
	};

	handleCloseModal = () => {
		this.props.handleCloseModal()
		this.setState({ newUser: false });
	}

	handleCancelNewUser = () => {
		this.setState({ newUser: false });
	}

	handleCreateNewUser = (e) => {
		e.preventDefault();

		firebase.auth().createUserWithEmailAndPassword(this.state.email, this.state.password).catch((e) => {
			if (e.code === "auth/email-already-in-use") {
				this.setState({ error: "That email is already in use." });
			} else if (e.code === "auth/invalid-email") {
				this.setState({ error: "Incorrect username or password." });
			} else {
				this.setState({ error: e.message });
			}
		});
	}

	handleGoogleLogin = () => {
		startLogin();
	}

	render() {
		return (
			<Modal
				ariaHideApp={false}
				onRequestClose={this.handleCloseModal}
				isOpen={!!this.props.modalOpen}
				contentLabel="Login"
				closeTimeoutMS={200}
				className="modal"
			>
				{this.state.newUser ?
					<form className="form" onSubmit={this.handleCreateNewUser}>
					{this.state.error && <p className="modal__error">{this.state.error}</p>}
					<input
						className="text-input"
						type="text"
						placeholder="Email"
						autoFocus
						value={this.state.email}
						onChange={this.onEmailChange}
					/>
					<input
						className="text-input"
						placeholder="Password"
						type="password"
						value={this.state.password}
						onChange={this.onPasswordChange}
					></input>
					<div className="modal-button">
						<button className="button">Create</button>
					</div>
					<div>
						<button className="button" onClick={this.handleCancelNewUser}>Cancel</button>	
					</div>
					<div className="modal-button__margin-top">
						<button className="button" onClick={this.handleCloseModal}>Close</button>
					</div>
					</form>
					:
					<div>
						<form className="form" onSubmit={this.handleLogin}>
							{this.state.error && <p className="modal__error">{this.state.error}</p>}
							<input
								className="text-input"
								type="text"
								placeholder="Email"
								autoFocus
								value={this.state.email}
								onChange={this.onEmailChange}
							/>
							<input
								className="text-input"
								placeholder="Password"
								type="password"
								value={this.state.password}
								onChange={this.onPasswordChange}
							></input>
							<div className="modal-button">
								<div className="modal-button__group">
									<button className="button">Sign In</button>
									<button className="button" onClick={this.handleStartNewUser}>New User</button>
								</div>
								<div className="modal-google" onClick={(() => this.props.startGoogleLogin())}>Google Login</div>
							</div>
						</form>
						<div>
							<button className="button" onClick={this.handleCloseModal}>Close</button>
						</div>
					</div>
				}
      </Modal>
		)
	}
}

const mapDispatchToProps = (dispatch) => ({
	startGoogleLogin: () => dispatch(startGoogleLogin())
});

export default connect(undefined, mapDispatchToProps)(LoginModal);