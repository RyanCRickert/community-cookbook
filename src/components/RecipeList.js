import React from "react";
import { connect } from "react-redux";
import RecipeListItem from "./RecipeListItem";
import SelectRecipes from "../selectors/recipes";

export const RecipeList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Recipes</div>
      <div className="show-for-larger">Recipe</div>
      <div className="show-for-larger">Cooking Time</div>
    </div>
    <div className="list-body">
  {
    props.recipes.length === 0 ? (
      <div className="list-item list-item--message">
        <span>No recipes found</span>
      </div>
    ) : (
      props.recipes.map((recipe) => (
      <RecipeListItem key={recipe.id} {...recipe}/>
    ))
    )
  }
    </div>
  </div>
);

const mapStateToProps = (state) => {
  return {
    recipes: SelectRecipes(state.recipes, state.filters)
  };
}

export default connect(mapStateToProps)(RecipeList);