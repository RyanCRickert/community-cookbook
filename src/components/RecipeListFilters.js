import React from "react";
import { connect } from "react-redux";
import { setNameFilter, sortByName, sortByCookTime, sortByDate, sortByFeeds } from "../actions/filters";

export class RecipeListFilters extends React.Component {
  onNameChange = (e) => {
    this.props.setNameFilter(e.target.value);
  };

  onSortChange = (e) => {
    if (e.target.value === "date") {
      this.props.sortByDate();
    } else if (e.target.value === "cookTime") {
      this.props.sortByCookTime();
    } else if (e.target.value === "name") {
      this.props.sortByName();
    } else if (e.target.value === "feeds") {
      this.props.sortByFeeds();
    }
  };

  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              type="text"
              className="text-input"
              placeholder="Search recipes"
              value={this.props.filters.name}
              onChange={this.onNameChange}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={this.onSortChange}
            >
              <option value="cookTime">Cook Time</option>
              <option value="date">Date</option>
              <option value="name">Name</option>
              <option value="feeds">Feeds</option>
            </select>
          </div>
        </div>
      </div>
    )
  }
};

const mapStateToProps = (state) => ({
  filters: state.filters
});

const mapDispatchToProps = (dispatch) => ({
  setNameFilter: (name) => dispatch(setNameFilter(name)),
  sortByDate: () => dispatch(sortByDate()),
  sortByName: () => dispatch(sortByName()),
  sortByCookTime: () => dispatch(sortByCookTime()),
  sortByFeeds: () => dispatch(sortByFeeds())
});

export default connect(mapStateToProps, mapDispatchToProps)(RecipeListFilters);