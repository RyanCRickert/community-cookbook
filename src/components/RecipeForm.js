import React from "react";
import moment from "moment";
import IngredientListItem from "./IngredientListItem";

export default class RecipeForm extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			name: props.recipe ? props.recipe.name : "",
			instructions: props.recipe ? props.recipe.instructions : "",
			ingredients: props.recipe ? props.recipe.ingredients : [],
			cookTime: props.recipe ? props.recipe.cookTime : "",
			createdAt: props.recipe ? moment(props.recipe.createdAt) : moment(),
			error: ""
		}
	}

	onNameChange = (e) => {
		const name = e.target.value;
		this.setState(() => ({ name }));
	};

	onCookChange = (e) => {
		const cookTime = e.target.value;
		this.setState(() => ({ cookTime }));
	};

	onInstructionsChange = (e) => {
		const instructions = e.target.value;
		this.setState(() => ({ instructions }));
	};

	onDateChange = (createdAt) => {
		if (createdAt) {
			this.setState(() => ({ createdAt }));
		}
	};

	handleRemoveIngredient = (ingredientToRemove) => {
		this.setState((prevState) => ({
			ingredients: prevState.ingredients.filter((ingredient) => ingredientToRemove !== ingredient)
		}));
	};

	handleAddIngredient = (e) => {
		e.preventDefault();

		const ingredient = e.target.elements.ingredient.value.trim();
		let error = false;

		if (!e.target.elements.ingredient.value) {
			this.setState({ error: "Please enter a valid ingredient" });
			error = true;
		} else if (this.state.ingredients.indexOf(ingredient) > -1) {
			this.setState({ error: "This ingredient already exists" });
			error = true;
		}

		if (!error) {
			e.target.elements.ingredient.value = "";
			this.setState((prevState) => ({ ingredients: prevState.ingredients.concat(ingredient) }));
			this.setState({ error: "" });
		}


	};

	onSubmit = (e) => {
		e.preventDefault();

		if (!this.state.name || !this.state.ingredients || !this.state.cookTime) {
			this.setState(() => ({ error: "Please enter a name, an ingredient and a cooking time." }));
		} else {
			this.setState(() => ({ error: "" }));
			this.props.onSubmit({
				name: this.state.name,
				ingredients: this.state.ingredients,
				createdAt: this.state.createdAt.valueOf(),
				instructions: this.state.instructions,
				cookTime: this.state.cookTime
			});
		}
	};

	render() {
		return (
			<div className="form">
				{this.state.error && <p className="form__error">{this.state.error}</p>}
				<input
					className="text-input"
					type="text"
					placeholder="Name of Dish"
					autoFocus
					value={this.state.name}
					onChange={this.onNameChange}
				/>
				<input
					className="text-input"
					type="text"
					placeholder="Cooking Time"
					value={this.state.cookTime}
					onChange={this.onCookChange}
				/>
				<form onSubmit={this.handleAddIngredient}>
					<input
						className="text-input__ingredient"
						type="text"
						name="ingredient"
						placeholder="Add Ingredient"
						value={this.state.ingredient}
					/>
					<button className="button text-input__button">Add Ingredient</button>
				</form>
				{this.state.ingredients.length === 0 ?
					<p>*Please add at least one ingredient*</p> :
					this.state.ingredients.map((ingredient) => (
						<IngredientListItem
							key={ingredient}
							name={ingredient}
							handleRemoveIngredient={this.handleRemoveIngredient}
						/>
					)
					)}
				<form className="form" onSubmit={this.onSubmit}>
					<textarea
						className="text-area"
						placeholder="Cooking instructions"
						value={this.state.instructions}
						onChange={this.onInstructionsChange}
					></textarea>
					<div>
						<button className="button">{this.props.actionType} Recipe</button>
					</div>
				</form>
			</div>
		)
	}
}