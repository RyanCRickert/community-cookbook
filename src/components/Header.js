import React from "react";
import { Link } from "react-router-dom";
import { connect } from "react-redux";
import { startLogout } from "../actions/auth";
import { firebase } from "../firebase/firebase";
import LoginModal from "./LoginModal";

export class Header extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      loggedIn: null,
      modalOpen: false
    }

    this.handleOpenModal = this.handleOpenModal.bind(this);
    this.handleCloseModal = this.handleCloseModal.bind(this);
  }

  componentDidMount() {
    firebase.auth().onAuthStateChanged((user) => {
      if (user) {
        this.setState(() => ({ loggedIn: true }));
      } else {
        this.setState(() => ({ loggedIn: false }));
      }
    })
  }

  handleOpenModal() {
    this.setState(() => ({ modalOpen: true }));
  }

  handleCloseModal() {
    this.setState(() => ({ modalOpen: false }));
  }

  render() {
    return (
      <header className="header">
        <div className="content-container">
          <div className="header-main">
            <h1 className="header-title">Community Cookbook</h1>
            <h3 className="header-subtitle">A place for sharing</h3>
          </div>
            {this.state.loggedIn ?
              <div className="header-log"><Link className="button header-button" to="/add">New Recipe</Link><button className="button header-button" onClick={this.props.startLogout}>Logout</button></div> :
              <div className="header-log"><LoginModal modalOpen={this.state.modalOpen} handleCloseModal={this.handleCloseModal}/><button className="button header-button header-button__login" onClick={this.handleOpenModal}>Login</button></div>}
        </div>
      </header>
    )
  }

};

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header)