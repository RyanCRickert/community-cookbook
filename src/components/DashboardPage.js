import React from "react";
import RecipeList from "./RecipeList";
import RecipeListFilters from "./RecipeListFilters";
import { Link } from "react-router-dom";

export default () => (
  <div>
    <h1><Link to="/add">Community Cookbook</Link></h1>
    <h3>A place for sharing</h3>
    <RecipeListFilters />
    <RecipeList />
  </div>
);