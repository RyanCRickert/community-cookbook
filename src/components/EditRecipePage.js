import React from "react";
import { connect } from "react-redux";
import RecipeForm from "./RecipeForm";
import { startEditRecipe, startRemoveRecipe } from "../actions/recipes";
import RemoveRecipeModal from "./RemoveRecipeModal";

export class EditRecipePage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      modalOpen: false
    }

    this.handleOpenModal = this.handleOpenModal.bind(this);
    this.handleCloseModal = this.handleCloseModal.bind(this);
  }
  onSubmit = (recipe) => {
    this.props.startEditRecipe(this.props.recipe.id, recipe);
    this.props.history.push("/");
  };

  onRemove = () => {
    this.props.startRemoveRecipe({id : this.props.recipe.id});
    this.props.history.push("/");
  }

  handleOpenModal() {
    this.setState(() => ({ modalOpen: true }));
  }

  handleCloseModal() {
    this.setState(() => ({ modalOpen: false }));
  }
    
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Recipe</h1>
          </div>
        </div>
          <div className="content-container">
            <RecipeForm
              actionType={"Edit"}
              recipe={this.props.recipe}
              onSubmit={this.onSubmit}
            />
          <button className="button button--secondary" onClick={this.handleOpenModal}>Remove Recipe</button>
          <RemoveRecipeModal
          modalOpen={this.state.modalOpen}
          handleCloseModal={this.handleCloseModal}
          onRemove={this.onRemove}
          recipeName={this.props.recipe.name}
          />
          </div>
        </div>
    )}
}

const mapStateToProps = (state, props) => ({
  recipe: state.recipes.find((recipe) => recipe.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  startEditRecipe: (id, recipe) => dispatch(startEditRecipe(id, recipe)),
  startRemoveRecipe: (data) => dispatch(startRemoveRecipe(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditRecipePage);