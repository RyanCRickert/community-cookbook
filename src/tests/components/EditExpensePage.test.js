import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import { EditExpensePage } from "../../components/EditExpensePage";
import expenses from "../fixtures/expenses";

let startEditExpense, history, wrapper, startRemoveExpense;

beforeEach(() => {
  startEditExpense = jest.fn();
  startRemoveExpense = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditExpensePage
      startEditExpense={startEditExpense}
      startRemoveExpense={startRemoveExpense}
      history={history}
      expense={expenses[0]}
    />
  );
});

test("should render a snapshot", () => {
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("should handle editExpense", () => {
  expect(toJSON(wrapper)).toMatchSnapshot();
  wrapper.find("ExpenseForm").prop("onSubmit")(expenses[0]);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(startEditExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test("should handle removeExpense", () => {
  expect(toJSON(wrapper)).toMatchSnapshot();
  wrapper.find("button").prop("onClick")(expenses[0]);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(startRemoveExpense).toHaveBeenLastCalledWith({
    id: expenses[0].id
  })
  expect(toJSON(wrapper)).toMatchSnapshot();
});