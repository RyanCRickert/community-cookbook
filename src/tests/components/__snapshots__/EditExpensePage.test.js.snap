// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle editExpense 1`] = `
<div>
  <ExpenseForm
    actionType="Edit"
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "0",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;

exports[`should handle editExpense 2`] = `
<div>
  <ExpenseForm
    actionType="Edit"
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "0",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;

exports[`should handle removeExpense 1`] = `
<div>
  <ExpenseForm
    actionType="Edit"
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "0",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;

exports[`should handle removeExpense 2`] = `
<div>
  <ExpenseForm
    actionType="Edit"
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "0",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;

exports[`should render a snapshot 1`] = `
<div>
  <ExpenseForm
    actionType="Edit"
    expense={
      Object {
        "amount": 195,
        "createdAt": 0,
        "description": "Gum",
        "id": "0",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
